### DISCUSSION TITLE
What may a "Parameter _fe_problem_base is NULL" means?

### DISCUSSION THREAD

**Discussion by WayneX-Wong** [idaholab/moose#24182] [Q&A General] [answered]
Hi all,

I'm trying to modify a MOOSE-based app to have my custom ability. The compilation process went smoothly, while when I run, an error raised "Parameter _fe_problem_base is NULL". What could be the reason for this error?

I inherit my custom UserObject from [TransientInterface](https://mooseframework.inl.gov/docs/doxygen/moose/TransientInterface_8h.html) to use `_t` to get current time, where I speculate that the problem may be. My custom UserObject indeed inherited from FEProblemBase, so I don't know exactly what the problem is.

Please let me know if more information about my modification needed. And any suggestions will be appreciated!

**Comment by GiudGiud** [comment]
Hello

It s very odd that your user object inherits from FEProblemBase and user objects. 
So that might be challenging. 
What are you trying to do?

The parameter error is not hard to fix I think. You need to find where it's retrieving the _fe_problem_base pointer parameter, and either change that code OR set the parameter like this:
```
params.set<FEProblemBase *>("_fe_problem_base") = &_fe_problem;
```
before retrieving it

Guillaume

**Reply by WayneX-Wong** [reply]
Thanks for your reply Guillaume. I will try to fix this error according to your suggestion.

> What are you trying to do?

I'm trying to realize a time dependent InputParameter in my user object, so I naturally thought of using `addParam<FunctionName>` to get a ParsedFunction, which need <Real> time and <Point> as input, so
> I inherit my custom UserObject from TransientInterface to use _t to get current time.

**Reply by GiudGiud** [reply]
Inheriting from transient interface seems fine. 

What does your object inherit from currently? Maybe a derived (General, Element etc) user object class makes sense

**Reply by WayneX-Wong** [reply]
I'm using OpenMC wrapped in [Cardinal](https://cardinal.cels.anl.gov/index.html), and try to make some modifications on [OpenMCCellAverageProblem](https://cardinal.cels.anl.gov/source/problems/OpenMCCellAverageProblem.html), which I think is inherited from [OpenMCProblemBase](https://cardinal.cels.anl.gov/doxygen/html/_open_m_c_problem_base_8h.html) and then [ExternalProblem](https://mooseframework.inl.gov/docs/doxygen/moose/ExternalProblem_8h_source.html), and some other MOOSE based class. So I think this issue may be more related to MOOSE and didn't ask on Cardinal's forum.

**Reply by GiudGiud** [reply]
ok be mindful that in Cardinal the time is not usually physical time. It might tied to openmc simulation parameters, I dont recall

**Reply by WayneX-Wong** [reply]
OK, Thanks again for your patience. I'll have another try

**Reply by aprilnovak** [reply]
Hi @WayneX-Wong - can you explain a bit more what you are trying to do? I might be able to give some ideas on how to accomplish specific things in Cardinal.

**Reply by WayneX-Wong** [reply]
Hello @aprilnovak , still that question I asked on Cardinal's forum about [CSG modeling parameter](https://github.com/neams-th-coe/cardinal/discussions/647), and accroding to your suggestion, which can be realized by fetching `openmc::model::surfaces` and then modify it in `OpenMCCellAverageProblem::setupProblem`.

Now I wanna make some further modification, to get time dependent geometry parameters in OpenMCCellAverageProblem's InputParameter, that's why I raised the above question.

I'm really sorry that my code was modified on a public LAN computer in the laboratory, so it is not easy to take out (request needed to be submitted). I will showcase my code if more details are needed, while it may take some time to request.

**Reply by aprilnovak** [reply]
What time-dependent information do you want to pass into OpenMC? You can access information in the form of postprocessors (scalar values), auxvariables (mesh-based data), or userobjects (something custom) from inside `OpenMCCellAverageProblem` (we do all three of these at the moment).

If I had to guess, you have a vector of numbers which represent different coefficients of the openMC CSG surfaces? Like, an array of length N where each entry is the new diameter of a pincell?

**Reply by WayneX-Wong** [reply]
> If I had to guess, you have a vector of numbers which represent different coefficients of the openMC CSG surfaces? Like, an array of length N where each entry is the new diameter of a pincell?

Exactly is. To be more specific, I try to modify a z0 of a ZPlane surface after each transient time step. What I thought is to get a ParsedFunction which describe the change of z0 wrt to time, then pass it into `OpenMCCellAverageProblem` so that I can change `openmc::model::surfaces` according to that function.

**Reply by aprilnovak** [reply]
All of the methods you can call inside of `OpenMCCellAverageProblem` are derived from `FEProblem` - if you go to the [Doxygen page](https://mooseframework.inl.gov/docs/doxygen/moose/classFEProblem.html), you'll see there is a method called `getFunction`. I think you should be able to use that to access your function with those values.

**Reply by WayneX-Wong** [reply]
> you'll see there is a method called getFunction

That's what I did. And the compilation process succeed. Only problem is 

> when I run, an error raised "Parameter _fe_problem_base is NULL".

And that's why I started this discussion.

**Reply by aprilnovak** [reply]
Ah, sorry for not catching that. I think you need to have `OpenMCCellAverageProblem` inherit from [FunctionInterface](https://mooseframework.inl.gov/docs/doxygen/moose/classFunctionInterface.html). I think you should be able to follow the pattern for how `OpenMCProblemBase` derives from `PostprocessorInterface`.

**Reply by WayneX-Wong** [reply]
Sorry for the late reply. I still have above doubts. I inherited `OpenMCProblemBase` from `FunctionInterface` and `TransientInterface`, completely following the way of inheriting `PostprocessorInterface`. There are no issues with the compilation process, but the compiled Cardinal-opt still raises an error of 

> 'Parameter _fe_problem_base is NULL'

 during runtime. I found the error raised at line24 of [FunctionInterface.C](https://mooseframework.inl.gov/docs/doxygen/moose/FunctionInterface_8C_source.html). Compared to `PostprocessorInterface`, `FunctionInterface` and `TransientInterface` have different ways of constructors definitions. I think the problem may lie here, but I cannot clearly tell.


Here are my specific modifications:

1. In `OpenMCProblemBase. h`, include `"FunctionInterface. h"` and `"TransientInterface. h"`

2. Initialize `FunctionInterface(this)` and `TransientInterface(this)` in the constructor of the `OpenMCProblemBase` in `OpenMCProblemBase. C`

3. Forward declare the class `Function` in `OpenMCCellAverageProblem. h` and declare a `Function` object member `_function` in `OpenMCCellAverageProblem` to receive functions from the input deck.

4. In `OpenMCCellAverageProblem. C`, include `"openmc/surface. h"` and `"Function. h"`, add `params. addRequiredParam<FunctionName>` in `validParams()` to receive the function name from the input deck, and initialize in the `OpenMCCellAverageProblem` constructor `_ function (getFunctionByName())`, finally used in `setupProblem` using `_ function ->value()` to modify geometry.


I've tried this also

> params.set<FEProblemBase *>("_fe_problem_base") = &_fe_problem;

But it gives a "_fe_problem was not declared in this scope" during compilation.

**Reply by GiudGiud** [reply]
so in the transient interface you have
```
TransientInterface::TransientInterface(const MooseObject * moose_object)
  : _ti_params(moose_object->parameters()),
    _ti_feproblem(*_ti_params.getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")),
```

this is likely how it s failing. 
Whatever mooseobject you are using to build the transient interface (in your constructor: you must have TransientInterface(SOMETHING) somewhere already) needs to have the fe_problem_base parameter already set.
This is usually done by the Action constructing the object. This is an intermediate layer between the input file and the creation of objects. 

Maybe this is not the way though. Like we need to re-consider the design
TransientInterface is meant for objects, not for problems. 
And having a user object be a Problem is odd as well.

**Reply by aprilnovak** [reply, marked as answer]
Hi @WayneX-Wong the actual issue is that the Problem classes in MOOSE are constructed before pretty much any other objects (except the `[Mesh]`). So within the constructor for `OpenMCCellAverageProblem`, functions, postprocessors, and user objects don't exist yet. So you need to fetch the function from inside `OpenMCCellAverageProblem::initialSetup`, which you can think of as an extended part of the constructor which we call after all the other objects in your input file have been constructed. My apologies for missing that detail!

To double-check everything, I implemented all the changes you would need as an example here: https://github.com/neams-th-coe/cardinal/pull/671 and confirmed that this works properly.

**Reply by WayneX-Wong** [reply]
@aprilnovak @GiudGiud I would like to thank you again for your patience and instruction! This solves my problem very well.
Guess I still got to improve my programming skill. Again, thanks so much for your guidance！

**Reply by aprilnovak** [reply]
Glad it worked! And no problem - the object order-of-construction requires a bit more advanced knowledge of MOOSE (and ideally, that error thrown by MOOSE would be less cryptic). Perhaps @GiudGiud has an idea on how to improve that error

**Reply by GiudGiud** [reply]
It's tough here it is a getParam that is failing, not a getFunction or a getUserObject which tend to be the errors for order-of-construction related issues.

We already tell the user on a failing getParam which parameter they have defined.
Maybe we could point to Debug/show_actions ? I feel like it would have been instructive here

I already raised this issue to improve this
https://github.com/idaholab/moose/issues/23807
